src/pybind/mgr/cephadm/module.py

```diff
@@ -1840,10 +1859,11 @@ def _check_daemons(self):
                 self._create_daemon(dd.daemon_type, dd.daemon_id,
                                     dd.hostname, reconfig=True)
 
-        if grafanas:
-            self.grafana_service.daemon_check_post(grafanas)
-        if iscsi_daemons:
-            self.iscsi_service.daemon_check_post(iscsi_daemons)
+        # perform post actions for daemons like checking or setting configs
+        for v in post_actions.values():
+            func = getattr(v.get('service'), 'daemon_check_post')
+            if func:
+                func(v.get('daemons'))
```

Comment:
```suggestion
        # perform post actions for daemons like checking or setting configs
        for whatever in post_actions.values():
            try:
                whatever['service'].daemon_check_post(whatever['daemons'])
            except:
```
... but I still find this code hard to understand.

BTW, as far as I can remember, the `shutdown` hook wasn't being run because of some leaks/segfaults in the Finisher thread... Has this changed recently?

Reply:
Should be easier to read..

https://github.com/ceph/ceph/pull/35194/files#diff-b2eb7e15b64fdea42d7c8afdab01bcbcR1860